software/contrib/noddy_holder.py

```diff
@@ -0,0 +1,138 @@
+from europi import *
+import machine
+from europi_script import EuroPiScript
+from time import ticks_diff, ticks_ms
+
+'''
+NoddyHolder
+author: Sean Bechhofer (github.com/seanbechhofer)
+date: 2022-03-26
+labels: sample & hold, trigger & hold
+
+Two channels of sample and hold/trigger and hold. 
+
+Digital in: Gate
+Analog in: CV
+
+Output 1: Gate
+Output 2: Sample and Hold based on Gate and CV
+Output 3: Trigger and Hold based on Gate and CV
+
+Output 4: Inverted Gate
+Output 5: Sample and Hold based on Inverted Gate and CV
+Output 6: Trigger and Hold based on Inverted Gate and CV
+
+digital_in: gate
+analog_in: cv
+
+knob_1: Not used
+knob_2: Not used
+
+button_1: Not used
+button_2: Not used
+
+output_1: gate 
+output_2: s&h
+output_3: t&h
+output_4: inverted gate
+output_5: s&h using inverted gate
+output_6: t&h using inverted gate
+
+'''
+# Version number
+
+VERSION="1.0"
+
+# Overclock the Pico for improved performance.
+machine.freq(250_000_000)
+
+class NoddyHolder(EuroPiScript):
+    def __init__(self):
+        self.gate = False
+        # Keep track of values for display, S&H and T&H for each
+        # channel
+        self.channel_1 = [0,-1]
+        self.channel_2 = [0,-1]
+        
+        # Triggered when din goes HIGH.
+        @din.handler
+        def dinTrigger():
+            self.gate = True
+            # Sample input
+            sample = ain.read_voltage()
+            # Set mirrored gate HIGH
+            cv1.on()
+            # Set inverse gate LOW
+            cv4.off()
+
+            # Set S&H to sample
+            cv2.voltage(sample)
+            self.channel_1[0] = sample
+            # Channel 1 is now tracking
+            
+            self.channel_1[1] = -1
+            # Set T&H of inverted gate to sample
+            cv6.voltage(sample)
+            self.channel_2[1] = sample
+            self.update_screen()
```

Comment:
Typically, for performance reasons,  it isn't recommended to do this much work inside of an interrupt handler. Instead the handler would just set a flag or make quick adjustments to a couple variables and then the main loop would check those variables an act on them.

In addition, by doing screen updates here you are tying your screen refresh rate to the clock cycle, but these are separate concerns. For example, if you were using knobs and buttons, you might want the user's interactions with them to update the screen more frequently than a slow clock cycle.

All that said, this script seems to do just fine in this configuration, since it's not using the controls, even with a fast clock. So I guess we don't need any immediate changes here, but I just want you to be aware of these concerns for any future scripts you may write.

Reply:
I'd thought about using something like the bar display on the Attractor, but decided that the display would get a bit too buy. As I don't include detailed information about the state of the tracked CV, screen refresh at clock rate works ok.